tikv/raftstore/snap_manager.go

```diff
@@ -0,0 +1,355 @@
+}
+
+func (sm *SnapManager) GetSnapshotForReceiving(snapKey SnapKey, data []byte) (Snapshot, error) {
+	snapshotData := new(rspb.RaftSnapshotData)
+	err := snapshotData.Unmarshal(data)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	return NewSnapForReceiving(sm.base, snapKey, snapshotData.Meta, sm.snapSize, sm, sm.limiter)
+}
+
+func (sm *SnapManager) GetSnapshotForApplying(snapKey SnapKey) (Snapshot, error) {
+	snap, err := NewSnapForApplying(sm.base, snapKey, sm.snapSize, sm)
+	if err != nil {
+		return nil, err
+	}
+	if !snap.Exists() {
+		return nil, errors.Errorf("snapshot of %s not exists", snapKey)
+	}
+	return snap, nil
+}
+
+func (sm *SnapManager) Register(key SnapKey, entry SnapEntry) {
+	log.Debugf("register key:%s, entry:%d", key, entry)
+	sm.registryLock.Lock()
+	defer sm.registryLock.Unlock()
+	entries, ok := sm.registry[key]
+	if ok {
+		for _, e := range entries {
+			if e == entry {
+				log.Warnf("%s is registered more than 1 time", key)
+				return
+			}
+		}
+	}
+	entries = append(entries, entry)
+	sm.registry[key] = entries
+	notifyStats(sm.ch)
+}
+
+func (sm *SnapManager) Deregister(key SnapKey, entry SnapEntry) {
+	log.Debugf("deregister key:%s, entry:%s", key, entry)
+	sm.registryLock.Lock()
+	defer sm.registryLock.Unlock()
+	var handled bool
+	entries, ok := sm.registry[key]
+	if ok {
+		for i, e := range entries {
+			if e == entry {
+				entries = append(entries[:i], entries[i+1:]...)
+				handled = true
+				break
+			}
+		}
+		if handled {
+			if len(entries) > 0 {
+				sm.registry[key] = entries
+			} else {
+				delete(sm.registry, key)
+			}
+			notifyStats(sm.ch)
+			return
+		}
+	}
+	log.Warnf("stale deregister key:%s, entry:%s", key, entry)
+}
+
+func (sm *SnapManager) Stats() SnapStats {
+	sm.registryLock.RLock()
+	defer sm.registryLock.RUnlock()
+	var sendingCount, receivingCount int
+	for _, entries := range sm.registry {
+		var isSending, isReceiving bool
+		for _, entry := range entries {
+			switch entry {
+			case SnapEntryGenerating, SnapEntrySending:
+				isSending = true
+			case SnapEntryReceiving, SnapEntryApplying:
+				isReceiving = true
+			}
+		}
+		if isSending {
+			sendingCount++
+		}
+		if isReceiving {
+			receivingCount++
+		}
+	}
+	return SnapStats{SendingCount: sendingCount, ReceivingCount: receivingCount}
+}
+
+func (sm *SnapManager) DeleteSnapshot(key SnapKey, snapshot Snapshot, checkEntry bool) bool {
+	sm.registryLock.Lock()
+	defer sm.registryLock.Unlock()
+	if checkEntry {
+		if e, ok := sm.registry[key]; ok {
+			if len(e) > 0 {
+				log.Infof("skip to delete %s since it's registered more than 1, registered entries %v",
+					snapshot.Path(), e)
+				return false
+			}
+		}
+	} else if _, ok := sm.registry[key]; ok {
+		log.Infof("skip to delete %s since it's registered.", snapshot.Path())
+		return false
+	}
+	snapshot.Delete()
+	return true
+}
+
+type SnapManagerBuilder struct {
+	maxTotalSize        uint64
```

Comment:
The max_write_bytes_per_sec is missing?

Reply:
badger doesn't support the limiter feature yet.

Reply:
Got it.